### PR TITLE
Charithjperera/ps4000a block mode

### PR DESCRIPTION
**Pull request by charithjp**
Adding support for the PS4444 and PS4824 in block mode, added resolution setting and copied over the PS5000a example and verified functionality on the PS4444.

### CODE REVIEW

File: picoscope/ps4000a.py [hunk 1]
```diff
@@ -278,28 +283,68 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
...
-        maxSampleTime = (((2 ** 32 - 1) - 4) / 2e7)
 
-        if sampleTimeS <= 12.5E-9:
-            timebase = math.floor(math.log(sampleTimeS * 8E7, 2))
-            timebase = max(timebase, 0)
-        else:
-            # Otherwise in range 2^32-1
-            if sampleTimeS > maxSampleTime:
-                sampleTimeS = maxSampleTime
+        if self.model in ('4828'):
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: I think this should be `if self.model in ('4828',):`, notice the comma after the `'4828'`

Comment (hmaarrfk) [reply picoscope/ps4000a.py]: or why not even `self.model == '4828'` ???

File: picoscope/ps4000a.py [hunk 2]
```diff
@@ -278,28 +283,68 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
...
+            if sampleTimeS <= 12.5E-9:
+                timebase = 0
+            else:
+                # Otherwise in range 2^32-1
+                if sampleTimeS > maxSampleTime:
+                    sampleTimeS = maxSampleTime
+
+                timebase = math.floor((sampleTimeS * 2e7) + 1)
+
+        elif self.model in ('4444'):
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: same as above

File: picoscope/ps4000a.py [hunk 3]
```diff
@@ -278,28 +283,68 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
...
+            else:
+                # Otherwise in range 2^32-1
+                if sampleTimeS > maxSampleTime:
+                    sampleTimeS = maxSampleTime
 
-            timebase = math.floor((sampleTimeS * 2e7) + 1)
+                timebase = math.floor((sampleTimeS * 5.0E7) + 2)
+
+        else: # The original case from non "A" series
+            print("Note - the model you are using may not be fully supported")
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: should be warning. Can you please use `stacklevel=2` https://docs.python.org/3.6/library/warnings.html#warnings.warn

File: picoscope/ps4000a.py [hunk 4]
```diff
@@ -278,28 +283,68 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
 
     def getTimeBaseNum(self, sampleTimeS):
         """Return sample time in seconds to timebase as int for API calls."""
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: This docstring makes no sense..... Can you think of a better one?

File: picoscope/ps4000a.py [hunk 5]
```diff
@@ -278,28 +283,68 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
...
+                    sampleTimeS = maxSampleTime
+
+                timebase = math.floor((sampleTimeS * 2e7) + 1)
 
         # is this cast needed?
         timebase = int(timebase)
         return timebase
 
     def getTimestepFromTimebase(self, timebase):
         """Return timebase to sampletime as seconds."""
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: this docstring makes no sense. Can you think of a better one (reading this code multiple years later makes these things more obvious).

File: picoscope/ps4000a.py [hunk 6]
```diff
@@ -278,28 +283,68 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
...
-        if timebase < 3:
-            dt = 2. ** timebase / 8e7
-        else:
-            dt = (timebase - 1) / 2e7
+        if self.model in ('4828'):
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: same as above.

File: picoscope/ps4000a.py [hunk 7]
```diff
@@ -278,28 +283,68 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
...
+            dt = (timebase + 1) / 8.0E7
+        elif self.model in ('4444'):
+            if timebase < 3:
+                dt = 2.5 ** timebase / 4.0E8
+            else:
+                dt = (timebase - 2) / 5.0E7
+
+        else: # The original case from non "A" series
+            print("Note - the model you are using may not be fully supported")
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: please use warnings.

File: picoscope/ps4000a.py [hunk 8]
```diff
@@ -313,11 +358,12 @@ def _lowLevelSetDataBuffer(self, channel, data, downSampleMode,
         segmentIndex is unused, but required by other versions of the API
         (eg PS5000a)
         """
+        dataPtr = data.ctypes.data_as(POINTER(c_int16))
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: why is this change necessary?

File: picoscope/ps4000a.py [hunk 9]
```diff
@@ -342,6 +388,35 @@ def _lowLevelGetValues(self, numSamples, startIndex, downSampleRatio,
...
+        m = self.lib.ps4000aSetDeviceResolution(
+            c_int16(self.handle),
+            c_enum(resolution))
+        self.checkResult(m)
+
+    def _lowLevelGetValuesBulk(self, numSamples, fromSegment, toSegment,
+                               downSampleRatio, downSampleMode, overflow):
+        """Copy data from several memory segments at once."""
+        overflowPoint = overflow.ctypes.data_as(POINTER(c_int16))
+        m = self.lib.ps4000aGetValuesBulk(
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: did you change anything but the formatting here? Is it possible for you to undo this change without too much work?


it is REALLY difficult to check the correctness of this kind of code, so I'm much more in favour of keeping ugly formatting than changing it.

**Comment by hmaarrfk** [comment]
Hey thank you for contributing back. I'm glad this is still useful to some even while picoscope released their own wrappers.

Could you please address the comments?

Before merging in, I will have to insist that PEP8 passes. This isn't so much because I care about style, but more because i am treating it as a static compiler of sorts because it is impossible to test this code in any automated fasion.

Also, be careful with `in` in python. The presence of the comma makes a big difference:

```
In [3]: 'abc' in ('abcd', )                                                                           
Out[3]: False

In [4]: 'abc' in ('abcd')                                                                             
Out[4]: True
```

File: picoscope/ps4000a.py (hunk 9 above)
Comment (charithjperera) [reply picoscope/ps4000a.py]: The ps4000a looked like it was a copy of ps4000 with a find and replace to refer to the ps4000a library. The low level call was modified to be more in line with the ps5000a library (uint16's modified to uint32's etc).

File: picoscope/ps4000a.py (hunk 8 above)
Comment (charithjperera) [reply picoscope/ps4000a.py]: Similar to before, the ps4000 library didn't do this properly so I mostly copied the implementation from ps5000a. Some of the variable names were changed along the way, but it became more consistent with a newer version.

File: picoscope/ps4000a.py (hunk 5 above)
Comment (charithjperera) [reply picoscope/ps4000a.py]: Okay, I improved the wording on that.

File: picoscope/ps4000a.py (hunk 4 above)
Comment (charithjperera) [reply picoscope/ps4000a.py]: ditto to below

File: picoscope/ps4000a.py (hunk 9 above)
Comment (hmaarrfk) [reply picoscope/ps4000a.py]: got it, thanks! keep it as is.

File: picoscope/ps4000a.py (hunk 1 above)
Comment (hmaarrfk) [reply picoscope/ps4000a.py]: Right, just let those fail. I've thought about making this more generated in the past, but I feel like there are enough small changes in the API and the dtypes that it isn't worth it.

File: picoscope/ps4000a.py [hunk 10]
```diff
@@ -277,29 +285,77 @@ def _lowLevelGetTimebase(self, tb, noSamples, oversample, segmentIndex):
...
+        if self.model == '4828':
+            dt = (timebase + 1) / 8.0E7
+        elif self.model == '4444':
+            if timebase < 3:
+                dt = 2.5 ** timebase / 4.0E8
+            else:
+                dt = (timebase - 2) / 5.0E7
+
+        else:  # The original case from non "A" series
+            warnings.warn(message="The model PS4000a you are using may not be"
```
Comment (hmaarrfk) [inline picoscope/ps4000a.py]: there needs to be a space after the "e" and before the quote.

If you feel like it, be as verbose as you want, you will thank me in the future ;)

File: picoscope/ps4000a.py (hunk 8 above)
Comment (hmaarrfk) [reply picoscope/ps4000a.py]: Thanks!

**Comment by hmaarrfk** [comment]
Just one small nit and I think we are ready to go. I'll cut a release after I merge this so you will be able to download it on PyPi.

**Comment by charithjperera** [comment]
Sweet, thanks Mark. I've made the style changes. Next step - I need to try and get those connected probes working on the 4444 and I'll send you a PR once I have that running :)

**Comment by hmaarrfk** [comment]
I'm confused, you never tested this code on hardware?

**Comment by charithjperera** [comment]
Thanks Mark, I missed the space between the two strings.
I tested on the 4444 with BNC adapters but it's the powered probes that I have not yet tried, they do something funky with the range value that's passed into the API.

**Comment by hmaarrfk** [comment]
Yeah, their API is kinda funky.

Good luck!
